src/libraries/System.Text.Json/src/System/Text/Json/Node/JsonValueOfT.cs

```diff
@@ -0,0 +1,375 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Diagnostics;
+using System.Diagnostics.CodeAnalysis;
+
+namespace System.Text.Json.Node
+{
+    [DebuggerDisplay("{ToJsonString(),nq}")]
+    [DebuggerTypeProxy(typeof(JsonValue<>.DebugView))]
+    internal sealed partial class JsonValue<TValue> : JsonValue
+    {
+        internal readonly TValue _value; // keep as a field for direct access to avoid copies
+
+        public JsonValue(TValue value, JsonNodeOptions? options = null) : base(options)
+        {
+            Debug.Assert(value != null);
+            Debug.Assert(!(value is JsonElement) || ((JsonElement)(object)value).ValueKind != JsonValueKind.Null);
+
+            if (value is JsonNode)
+            {
+                ThrowHelper.ThrowArgumentException_NodeValueNotAllowed(nameof(value));
+            }
+
+            _value = value;
+        }
+
+        public TValue Value
+        {
+            get
+            {
+                return _value;
+            }
+        }
+
+        public override T GetValue<[DynamicallyAccessedMembers(MembersAccessedOnRead)] T>()
+        {
+            Type returnType = typeof(T);
+
+            // If no conversion is needed, just return the raw value.
+            if (_value is T returnValue)
+            {
+                return returnValue;
+            }
+
+            if (_value is JsonElement jsonElement)
+            {
+                return ConvertJsonElement<T>();
```

Comment:
`ConvertJsonElement` and `TryConvertJsonElement` seem very similar. Could we instead call `TryConvertJsonElement` here, and throw the `IOE` if it returns false? Perhaps we could then delete `ConvertJsonElement`.

Reply:
Perhaps. The reason is since we delegate to `JsonElement` to obtain the values in `ConvertJsonElement` we leverage the existing validation and exception logic in `JsonElement`. `JsonElement` in some cases throws `FormatException` and in other cases `InvalidOperation` which we would need to duplicate if we combine.